Added implementation of linked vertex preprocessing

revel8n opened the pull request.

revel8n committed:
Added implementation of linked vertex preprocessing

MrSapps commented on Source/MgsLib/LibDG.cpp:
```diff
@@ -645,12 +645,53 @@ MGS_FUNC_IMPLEX(0x401A4F, LibGvInitDispEnv_401A4F, LIBDG_IMPL);
 {
-   
+    unsigned int indexFlags = 0;
+    BYTE* vertexIndices = (BYTE*)pKmdObj->indexOfs_3C;
```
indexOfs_3C is already BYTE* so I think this cast is redundant?

MrSapps commented on Source/MgsLib/LibDG.cpp:
```diff
@@ -645,12 +645,53 @@ MGS_FUNC_IMPLEX(0x401A4F, LibGvInitDispEnv_401A4F, LIBDG_IMPL);
+    for (int i = pKmdObj->field_4_numFaces * 4; i > 0; --i, ++vertexIndices)
+    {
+        const SVECTOR& vertexData = pKmdObj->vertOfs_38[*vertexIndices];
+        unsigned short linkedVertexIndex = vertexData.field_6_padding;
```
I would make locals const where possible

MrSapps commented on Source/MgsLib/LibDG.cpp:
```diff
@@ -645,12 +645,53 @@ MGS_FUNC_IMPLEX(0x401A4F, LibGvInitDispEnv_401A4F, LIBDG_IMPL);
+    }
+
+    // determine whether vertex index data has already been processed
+    bool processed = (0x80 & indexFlags) != 0;
```
ditto const

MrSapps commented on Source/MgsLib/LibDG.cpp:
```diff
@@ -645,12 +645,53 @@ MGS_FUNC_IMPLEX(0x401A4F, LibGvInitDispEnv_401A4F, LIBDG_IMPL);
+    SVECTOR* vertexData = pKmdObj->vertOfs_38;
+    for (int i = pKmdObj->numVerts_34; i > 0; --i, ++vertexData)
+    {
+        unsigned short linkedVertexIndex = vertexData->field_6_padding;
```
ditto const

MrSapps commented on Source/MgsLib/LibDG.cpp:
```diff
@@ -645,12 +645,53 @@ MGS_FUNC_IMPLEX(0x401A4F, LibGvInitDispEnv_401A4F, LIBDG_IMPL);
+                break;
+        }
+
+        int offset = (parentVertexIndices - (BYTE*)pParentObj->indexOfs_3C);
```
ditto const for these 3

MrSapps commented on Source/MgsLib/LibDG.cpp:
```diff
@@ -645,12 +645,53 @@ MGS_FUNC_IMPLEX(0x401A4F, LibGvInitDispEnv_401A4F, LIBDG_IMPL);
+        int faceVertexIndex = offset % 4;
+
+        // compute offset of linked poly data
+        vertexData->field_6_padding = (sizeof(POLY_GT4) * faceIndex) + (0xC * k0132_byte_650174[faceVertexIndex]) + 8;
```
Maybe add a comment for what the 0xC and + 8 is for?

MrSapps replied:
hmm, perhaps it might be better if we do something like

switch (faceVertexIndex)
{
case 1: vertexData->field_6_padding = (sizeof(POLY_GT4) * faceIndex) +offsetof(POLY_GT4::x0);
}

Or some such like that ?

revel8n replied:
Although probably clearer on what it is trying to do, i doubt the branching within the switch would be optimized away.

revel8n committed:
Updates to improve clarity of the code

- Made local variables const where appropriate
- Removed unnecessary casts
- Added additional comments to clarify data offset calculations

MrSapps commented on Source/MgsLib/LibDG.cpp:
```diff
@@ -645,12 +645,56 @@ MGS_FUNC_IMPLEX(0x401A4F, LibGvInitDispEnv_401A4F, LIBDG_IMPL);
-   
+    unsigned int indexFlags = 0;
+    BYTE* vertexIndices = pKmdObj->indexOfs_3C;
+    for (int i = pKmdObj->field_4_numFaces * 4; i > 0; --i, ++vertexIndices)
```
Even though the actual loops are backwards in the game I think we should use regular loops as its easier to understand.

MrSapps commented on Source/MgsLib/LibDG.cpp:
```diff
@@ -645,12 +645,56 @@ MGS_FUNC_IMPLEX(0x401A4F, LibGvInitDispEnv_401A4F, LIBDG_IMPL);
 
 void CC GV_kmd_zmd_file_handler_helper_4028C1(kmdObject* pKmdObj, kmdObject* pParentObj);
-MGS_FUNC_IMPLEX(0x4028C1, GV_kmd_zmd_file_handler_helper_4028C1, false); // TODO
+MGS_FUNC_IMPLEX(0x4028C1, GV_kmd_zmd_file_handler_helper_4028C1, LIBDG_IMPL); // TODO
```
MGS_FUNC_IMPLEX can probably be moved after the function now that it is implemented and the prototype/TODO removed.

MrSapps commented on Source/MgsLib/LibDG.cpp:
```diff
@@ -645,12 +645,56 @@ MGS_FUNC_IMPLEX(0x401A4F, LibGvInitDispEnv_401A4F, LIBDG_IMPL);
+        const SVECTOR& vertexData = pKmdObj->vertOfs_38[*vertexIndices];
+        const unsigned short linkedVertexIndex = vertexData.field_6_padding;
+
+        if (0xFFFF == linkedVertexIndex)
```
For single line statements the coding convention in the rest of the project is to till use braces. Also I think you should use linkedVertexIndex ==  0xFFFF instead of a yoda condition. Given linkedVertexIndex using = instead of == can't cause a bug or should generate a warning if not const :)

MrSapps commented on Source/MgsLib/LibDG.cpp:
```diff
@@ -645,12 +645,56 @@ MGS_FUNC_IMPLEX(0x401A4F, LibGvInitDispEnv_401A4F, LIBDG_IMPL);
+            continue;
+
+        const BYTE* parentVertexIndices = pParentObj->indexOfs_3C;
+        for (int j = pParentObj->field_4_numFaces * 4; j > 0; --j, ++parentVertexIndices)
```
Ditto for backwards loops

MrSapps commented on Source/MgsLib/LibDG.cpp:
```diff
@@ -645,12 +645,56 @@ MGS_FUNC_IMPLEX(0x401A4F, LibGvInitDispEnv_401A4F, LIBDG_IMPL);
+    {
+        const unsigned short linkedVertexIndex = vertexData->field_6_padding;
+
+        if (0xFFFF == linkedVertexIndex)
```
Ditto for yoda condition

MrSapps commented on Source/MgsLib/LibDG.cpp:
```diff
@@ -645,12 +645,56 @@ MGS_FUNC_IMPLEX(0x401A4F, LibGvInitDispEnv_401A4F, LIBDG_IMPL);
+        // 0xC == size of each transformed vertex section (rgb, code/p, xy, uv, clut/tpage/pad) within POLY_GT4
+        // 0x8 == start offset to xy data of first transformed vertex section within POLY_GT4
+        // k0132_byte_650174 provides the indexing order for referencing the transformed vertex sections
+        vertexData->field_6_padding = (sizeof(POLY_GT4) * faceIndex) + (0xC * k0132_byte_650174[faceVertexIndex]) + 8;
```
Maybe we should rename k0132_byte_650174 to kVertexIndexingOrder_650174? Also instead of 0xC I think that offsetof(POLY_GT4, x0) might be more clear? Or checking what the vert index order is and do offsetof(POLY_GT4, x0)/offsetof(POLY_GT4, x1)/offsetof(POLY_GT4, x2) etc.

revel8n replied:
i like this suggestion (keeping the original for reference):
```c++
// kVertexIndexingOrder_650174 provides the indexing order for referencing the transformed vertex sections
const char kVertexIndexingOrder_650174[4] = { 0, 1, 3, 2 };
// kVertexIndexingOffsets provides the offsets to the transformed vertex sections based on the indexing order
const char kVertexIndexingOffsets[4] = { offsetof(POLY_GT4, x0), offsetof(POLY_GT4, x1), offsetof(POLY_GT4, x3), offsetof(POLY_GT4, x2) };
```

MrSapps commented on Source/MgsLib/LibDG.cpp:
```diff
@@ -645,12 +645,56 @@ MGS_FUNC_IMPLEX(0x401A4F, LibGvInitDispEnv_401A4F, LIBDG_IMPL);
+    // determine whether vertex index data has already been processed
+    const bool processed = (0x80 & indexFlags) != 0;
+
+    if (processed)
```
Ditto single line statement braces

MrSapps commented on Source/MgsLib/LibDG.cpp:
```diff
@@ -645,12 +645,56 @@ MGS_FUNC_IMPLEX(0x401A4F, LibGvInitDispEnv_401A4F, LIBDG_IMPL);
+MGS_FUNC_IMPLEX(0x4028C1, GV_kmd_zmd_file_handler_helper_4028C1, LIBDG_IMPL); // TODO
 
 
 void CC GV_kmd_zmd_file_handler_helper_4028C1(kmdObject* pKmdObj, kmdObject* pParentObj)
```
I wonder if GV_kmd_zmd_file_handler_helper_4028C1 should now be renamed to like: GV_link_vertices_to_parent_4028C1 or something like that?

revel8n committed:
Clean up code to match conventions

- Renamed function to match its usage
- Removed prototype and moved patch macro
- Reversed loop order and used indexing
- Added array with explicit offsets instead of calculating from magic numbers
- Added braces where required by conventions
